rewrite-gradle/src/main/java/org/openrewrite/gradle/RemoveRedundantSecurityResolutionRules.java

```diff
@@ -0,0 +1,683 @@
+                          "(e.g., `CVE-2024-1234, GHSA-abcd-1234-efgh`) will match if any identifier matches. " +
+                          "Default pattern matches CVE identifiers (e.g., `CVE-2024-1234`) and GitHub Security " +
+                          "Advisory identifiers (e.g., `GHSA-xxxx-xxxx-xxxx`).",
+            example = "(CVE-\\d|GHSA-[a-z0-9])",
+            required = false)
+    @Nullable
+    String securityPattern;
+
+    @Override
+    public String getDisplayName() {
+        return "Remove redundant security resolution rules";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Remove `resolutionStrategy.eachDependency` rules that pin dependencies to versions that are already " +
+               "being managed by a platform/BOM to equal or newer versions. Only removes rules that have a security " +
+               "advisory identifier (CVE or GHSA) in the `because` clause, unless a custom pattern is specified.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return singleton("security");
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        Pattern compiledPattern = Pattern.compile(
+                securityPattern != null ? securityPattern : DEFAULT_SECURITY_PATTERN,
+                Pattern.CASE_INSENSITIVE
+        );
+
+        JavaIsoVisitor<ExecutionContext> removeRulesVisitor = new JavaIsoVisitor<ExecutionContext>() {
+            @Nullable
+            GradleProject gradleProject;
+            @Nullable
+            UUID currentSourceFileId;
+            boolean insideBuildscript;
+
+            private void maybeInitialize(JavaSourceFile sourceFile) {
+                // Reset state when visiting a new source file
+                if (currentSourceFileId != null && currentSourceFileId.equals(sourceFile.getId())) {
+                    return;
+                }
+                currentSourceFileId = sourceFile.getId();
+                gradleProject = sourceFile.getMarkers().findFirst(GradleProject.class).orElse(null);
+            }
+
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                JavaSourceFile sourceFile = getCursor().firstEnclosing(JavaSourceFile.class);
+                if (sourceFile != null) {
+                    maybeInitialize(sourceFile);
+                }
+
+                // Track when we enter/exit buildscript block
+                boolean wasInsideBuildscript = insideBuildscript;
+                if (METHOD_BUILDSCRIPT.equals(method.getSimpleName())) {
+                    insideBuildscript = true;
+                }
+
+                J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
+
+                // Restore state after visiting children
+                boolean currentlyInBuildscript = insideBuildscript;
+                insideBuildscript = wasInsideBuildscript;
```

Comment:
If you make `insideBuildscript` a field on the visitor class this can be as simple as
```java
if (METHOD_BUILDSCRIPT.equals(method.getSimpleName())) {
    insideBuildscript = true;
}
J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
insideBulidscript = false;
```

Reply:
done